File: services/agent-gateway/vss_agent_gateway/service.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Run lifecycle orchestration independent of HTTP and upstream protocol."""

from __future__ import annotations

import logging
import threading

from .artifacts import (
    ARTIFACT_PROTOCOL_VERSION,
    ArtifactStreamParser,
    strip_artifacts_from_value,
)
from .config import GatewayConfig
from .connectors.base import Connector, ConnectorError
from .contract import PROTOCOL_VERSION, CreateRunRequest
from .store import RunRecord, RunStore

LOGGER = logging.getLogger(__name__)


class GatewayService:
    def __init__(self, config: GatewayConfig, connector: Connector) -> None:
        self.config = config
        self.connector = connector
        self.store = RunStore(
            retention_seconds=config.run_retention_seconds,
            max_runs=config.max_runs,
            max_events_per_run=config.max_events_per_run,
            max_event_chars_per_run=config.max_event_chars_per_run,
        )

    def capabilities(self) -> dict[str, object]:
        capabilities: dict[str, object] = {
            "protocol_version": PROTOCOL_VERSION,
            "transport": "sse",
            "features": {
                "reconnect": True,
                "cancellation": True,
                "idempotent_run_creation": True,
                "interaction_responses": False,
                "artifacts": True,
            },
            "artifact_protocol": {
                "version": ARTIFACT_PROTOCOL_VERSION,
                "transport": "agent-text-envelope",
                "kinds": ["vss.search.results", "vss.alert.incidents"],
            },
            "connector": self.connector.capabilities,
            "event_types": [
                "run.started",
                "message.delta",
                "reasoning.delta",
                "tool.started",
                "tool.arguments.delta",
                "tool.requested",
                "tool.completed",
                "tool.failed",
                "artifact.created",
                "interaction.required",
                "run.completed",
                "run.failed",
                "run.cancelled",
            ],
            "limits": {
                "max_events_per_run": self.config.max_events_per_run,
                "max_event_chars_per_run": self.config.max_event_chars_per_run,
                "run_retention_seconds": self.config.run_retention_seconds,
            },
        }
        capabilities["vss"] = (
            self.config.vss_capabilities.public_summary()
            if self.config.vss_capabilities is not None
            else {"attached": False, "ready": False}
        )
        return capabilities

    def create_run(
        self,
        request: CreateRunRequest,
        *,
        idempotency_key: str | None,
    ) -> tuple[RunRecord, bool]:
        record, replayed = self.store.create(request, idempotency_key=idempotency_key)
        if replayed:
            return record, True
        record.append(
            "run.started",
            {
                "surface": request.surface,
                "connector_protocol": self.connector.protocol,
            },
        )
        worker = threading.Thread(
            target=self._run_worker,
            args=(record,),
            daemon=True,
            name=f"agent-gateway-{record.run_id[-8:]}",
        )
        worker.start()
        return record, False

    def _run_worker(self, record: RunRecord) -> None:
        artifact_parser = ArtifactStreamParser()

        def append(event_type: str, data: dict[str, object]) -> None:
            if event_type == "message.delta":
                delta = data.get("delta")
                if isinstance(delta, str):
                    for parsed in artifact_parser.feed(delta):
                        record.append(parsed.type, parsed.data)
                    return
            if event_type == "tool.completed":
                output = data.get("output")
                artifacts = artifact_parser.inspect_complete(output)
                if output is not None:
                    data = {**data, "output": strip_artifacts_from_value(output)}
                record.append(event_type, data)
                for parsed in artifacts:
                    record.append(parsed.type, parsed.data)
                return
            record.append(event_type, data)

        def flush_artifact_parser() -> None:
            for parsed in artifact_parser.finish():
                record.append(parsed.type, parsed.data)

        try:
            for event in self.connector.run(
                record.request,
                run_id=record.run_id,
                cancel_event=record.cancel_event,
            ):
                if record.cancel_event.is_set():
                    break
                if event.type.startswith("run."):
                    raise ConnectorError(
                        "connector emitted a reserved terminal event",
                        code="connector_contract_error",
                    )
                append(event.type, event.data)
            if record.cancel_event.is_set():
                flush_artifact_parser()
                self.store.finish(
                    record, "run.cancelled", {"reason": "client_cancelled"}
                )
            else:
                flush_artifact_parser()
                self.store.finish(record, "run.completed")
        except ConnectorError as error:
            flush_artifact_parser()
            if record.cancel_event.is_set():
                self.store.finish(
                    record, "run.cancelled", {"reason": "client_cancelled"}
                )
            else:
                self.store.finish(
                    record,
                    "run.failed",
                    {
                        "error": {
                            "code": error.code,
                            "message": str(error),
                            "retryable": error.retryable,
                        },
                    },
                )
        except Exception:
            flush_artifact_parser()
            LOGGER.exception("unexpected connector failure for run %s", record.run_id)
            if record.cancel_event.is_set():
                self.store.finish(
                    record, "run.cancelled", {"reason": "client_cancelled"}
                )
            else:
                self.store.finish(
                    record,
                    "run.failed",
                    {
                        "error": {
                            "code": "gateway_internal_error",
                            "message": "the gateway could not complete this run",
                            "retryable": False,
                        },
                    },
                )

    def cancel_run(self, run_id: str) -> RunRecord:
        record = self.store.get(run_id)
        if not record.terminal:
            record.cancel_event.set()
            try:
                self.connector.cancel(run_id)
            except Exception:
                LOGGER.exception("connector cancellation failed for run %s", run_id)
        return record

File: services/ui/apps/nv-metropolis-bp-vss-ui/utils/server/agentAdapter/service.ts
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  ARTIFACT_PROTOCOL_VERSION,
  ArtifactStreamParser,
  stripArtifactsFromValue,
} from "./artifacts";
import type { AgentAdapterConfig } from "./config";
import { type Connector, ConnectorError } from "./connectors/base";
import { LegacyChatConnector } from "./connectors/legacyChat";
import { OpenClawConnector } from "./connectors/openClaw";
import { ResponsesConnector } from "./connectors/responses";
import {
  PROTOCOL_VERSION,
  type CreateRunRequest,
  type JsonObject,
} from "./contract";
import { RunRecord, RunStore } from "./store";

const buildConnector = (config: AgentAdapterConfig): Connector => {
  if (config.backendProtocol === "openclaw-ws") {
    return new OpenClawConnector(config);
  }
  if (config.backendProtocol === "legacy-chat") {
    return new LegacyChatConnector(config);
  }
  return new ResponsesConnector(config);
};

export class AgentAdapterService {
  readonly store: RunStore;
  private readonly connector: Connector;

  constructor(readonly config: AgentAdapterConfig) {
    this.connector = buildConnector(config);
    // Reserve the thread-state ceiling up front so the independently managed
    // connector cache and run store cannot exceed the process-wide limit.
    this.store = new RunStore(
      config.runRetentionMs,
      config.maxRuns,
      config.maxEventsPerRun,
      config.maxEventCharsPerRun,
      config.maxRetainedChars - config.maxThreadStateChars
    );
  }

  capabilities(): JsonObject {
    return {
      protocol_version: PROTOCOL_VERSION,
      transport: "sse",
      features: {
        reconnect: true,
        cancellation: true,
        idempotent_run_creation: true,
        interaction_responses: false,
        artifacts: true,
      },
      artifact_protocol: {
        version: ARTIFACT_PROTOCOL_VERSION,
        transport: "connector-normalized",
        transports: [
          "openclaw-tool-result",
          "openclaw-managed-image",
          "vss-cli-completion",
          "responses-client-tool",
          "agent-tool-output",
          "agent-text-envelope",
        ],
        kinds: [
          "vss.search.results",
          "vss.alert.incidents",
          "vss.media.image",
        ],
      },
      connector: this.connector.capabilities,
      event_types: [
        "run.started",
        "message.delta",
        "reasoning.delta",
        "tool.started",
        "tool.arguments.delta",
        "tool.requested",
        "tool.completed",
        "tool.failed",
        "artifact.created",
        "interaction.required",
        "run.completed",
        "run.failed",
        "run.cancelled",
      ],
      limits: {
        max_events_per_run: this.config.maxEventsPerRun,
        max_event_chars_per_run: this.config.maxEventCharsPerRun,
        max_retained_chars: this.config.maxRetainedChars,
        run_retention_seconds: this.config.runRetentionMs / 1_000,
      },
    };
  }

  createRun(
    request: CreateRunRequest,
    idempotencyKey?: string
  ): { record: RunRecord; replayed: boolean } {
    const created = this.store.create(request, idempotencyKey);
    if (created.replayed) return created;
    created.record.append("run.started", {
      surface: request.surface,
      connector_protocol: this.connector.protocol,
    });
    void this.executeRun(created.record);
    return created;
  }

  private static appendParsedEvents(
    record: RunRecord,
    events: Array<{ type: string; data: JsonObject }>
  ): void {
    for (const event of events) record.append(event.type, event.data);
  }

  private static toolArtifactSource(data: JsonObject): unknown {
    if (data._artifact_source !== undefined) return data._artifact_source;
    if (data.output !== undefined) return data.output;
    return data.payload;
  }

  private static appendToolCompletion(
    record: RunRecord,
    parser: ArtifactStreamParser,
    rawData: JsonObject
  ): void {
    const data = { ...rawData };
    const artifacts = parser.inspectComplete(
      AgentAdapterService.toolArtifactSource(data)
    );
    delete data._artifact_source;
    if (data.output !== undefined) {
      data.output = stripArtifactsFromValue(data.output);
    } else if (data.payload !== undefined) {
      data.payload = stripArtifactsFromValue(data.payload);
    }
    record.append("tool.completed", data);
    AgentAdapterService.appendParsedEvents(record, artifacts);
  }

  private static appendConnectorEvent(
    record: RunRecord,
    parser: ArtifactStreamParser,
    type: string,
    rawData: JsonObject
  ): void {
    if (type === "message.delta" && typeof rawData.delta === "string") {
      AgentAdapterService.appendParsedEvents(
        record,
        parser.feed(rawData.delta)
      );
      return;
    }
    if (type === "artifact.source") {
      AgentAdapterService.appendParsedEvents(
        record,
        parser.inspectComplete(rawData.source)
      );
      return;
    }
    if (type === "tool.completed") {
      AgentAdapterService.appendToolCompletion(record, parser, rawData);
      return;
    }
    record.append(type, rawData);
  }

  private async executeRun(record: RunRecord): Promise<void> {
    const parser = new ArtifactStreamParser(true);
    const flush = (): void => {
      AgentAdapterService.appendParsedEvents(record, parser.finish());
    };

    try {
      for await (const event of this.connector.run(
        record.request,
        record.runId,
        record.abortController.signal
      )) {
        if (record.abortController.signal.aborted) break;
        if (event.type.startsWith("run.")) {
          throw new ConnectorError(
            "connector emitted a reserved terminal event",
            "connector_contract_error"
          );
        }
        AgentAdapterService.appendConnectorEvent(
          record,
          parser,
          event.type,
          event.data
        );
      }
      flush();
      if (record.abortController.signal.aborted) {
        this.store.finish(record, "run.cancelled", {
          reason: "client_cancelled",
        });
      } else {
        this.store.finish(record, "run.completed");
      }
    } catch (error) {
      flush();
      if (record.abortController.signal.aborted) {
        this.store.finish(record, "run.cancelled", {
          reason: "client_cancelled",
        });
      } else if (error instanceof ConnectorError) {
        this.store.finish(record, "run.failed", {
          error: {
            code: error.code,
            message: error.message,
            retryable: error.retryable,
          },
        });
      } else {
        console.error("Unexpected embedded agent adapter failure");
        this.store.finish(record, "run.failed", {
          error: {
            code: "adapter_internal_error",
            message: "the adapter could not complete this run",
            retryable: false,
          },
        });
      }
    }
  }

  async cancelRun(runId: string): Promise<RunRecord> {
    const record = this.store.get(runId);
    if (!record.terminal) {
      record.abortController.abort(new Error("client cancelled"));
      try {
        await this.connector.cancel(runId);
      } catch {
        console.error("Embedded agent connector cancellation failed");
      }
    }
    return record;
  }
}
